server/src/main/java/io/crate/types/CharacterType.java

```diff
@@ -0,0 +1,194 @@
+import io.crate.sql.tree.ColumnPolicy;
+import io.crate.sql.tree.ColumnType;
+import io.crate.sql.tree.Expression;
+
+public class CharacterType extends StringType {
+
+    public static final String NAME = "character";
+    public static final int ID = 28;
+    public static final CharacterType INSTANCE = new CharacterType();
+
+    public static CharacterType of(List<Integer> parameters) {
+        if (parameters.size() != 1) {
+            throw new IllegalArgumentException(
+                "The character type can only have a single parameter value, received: " +
+                    parameters.size()
+            );
+        }
+        return CharacterType.of(parameters.get(0));
+    }
+
+    public static CharacterType of(int lengthLimit) {
+        if (lengthLimit <= 0) {
+            throw new IllegalArgumentException(
+                "The character type length must be at least 1, received: " + lengthLimit);
+        }
+        return new CharacterType(lengthLimit);
+    }
+
+    private final int lengthLimit;
+
+    public CharacterType(StreamInput in) throws IOException {
+        if (in.getVersion().onOrAfter(Version.V_5_0_0)) {
+            lengthLimit = in.readInt();
+        } else {
+            lengthLimit = 1;
+        }
+    }
+
+    public CharacterType(int lengthLimit) {
+        this.lengthLimit = lengthLimit;
+    }
+
+    private CharacterType() {
+        this(1);
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    @Override
+    public int id() {
+        return ID;
+    }
+
+    @Override
+    public int lengthLimit() {
+        return lengthLimit;
+    }
+
+    @Override
+    public boolean unbound() {
+        return false;
+    }
+
+    @Override
+    public String valueForInsert(Object value) {
+        if (value == null) {
+            return null;
+        }
+        assert value instanceof String
+            : "valueForInsert must be called only on objects of String type";
+        var string = (String) value;
+        if (string.length() == lengthLimit) {
+            return string;
+        } else if (string.length() < lengthLimit) {
+            return pad(string);
+        } else {
+            if (isBlank(string, lengthLimit, string.length())) {
+                return string.substring(0, lengthLimit);
+            } else {
+                if (string.length() > 20) {
+                    string = string.substring(0, 20) + "...";
+                }
+                throw new IllegalArgumentException(
+                    "'" + string + "' is too long for the character type of length: " + lengthLimit);
+            }
+        }
+    }
+
+    private String pad(String value) {
+        char[] buffer = new char[lengthLimit];
+        System.arraycopy(value.toCharArray(), 0, buffer, 0, value.length());
+        for (int i = value.length(); i < lengthLimit; i++) {
+            buffer[i] = ' ';
+        }
+        return String.valueOf(buffer);
```

Comment:
There would also be a `padEnd` in `StringUtils`.  (and I think a regular string concat might even be faster, but didn't verify)

Reply:
Wasn't aware of this. I've run a quick benchmark, comparing all 3 variants with a string `foo` (3 chars) and padding to 10 chars with whitespaces:
```
Benchmark                           Mode  Cnt         Score         Error  Units
CharacterTypeTest.padConcatRepeat  thrpt   25  95202620,469 ± 3391085,804  ops/s
CharacterTypeTest.padCustom        thrpt   25  46869678,245 ±  933217,198  ops/s
CharacterTypeTest.padStringUtils   thrpt   25  35043663,237 ±  835865,710  ops/s
```

Whereas `padConcatRepeat` is implemented like this:
```"foo" + " ".repeat(7);```

So seems that concat + repeat wins, even a bit surprising as it uses a similar approach than my custom pad function.

Reply:
> Based on these results, shall we use concat + repeat inside StringUtils.padEnd as well?

I'd say yes.

> even a bit surprising as it uses a similar approach than my custom pad function.

I think the JDK added some intrinsics for the JIT/c2 compiler at some point, but I could be mixing things up.